### PR TITLE
Paradox Abilities

### PR DESCRIPTION
**Pull request by Koekenpann**
Added tests from pokeemerald expansion and some edge cases.

Tests maybe worth adding:
* Cloud Nine/Air Lock -> deactivation/no activation
* Switching Weather/Terrain -> deactivation
* Defense chosen in tie
* Speed modifier is 1.5x (6144/4096), not 1.3 (5325/4096) or vice versa

# Paradox Abilities

## Implemented Abilities
* Protosynthesis
* Quark Drive
* Orichalcum Pulse
* Hadron Engine

## Implemented Items
* Booster Energy
* Utility Umbrella (partially: interactions with Protosynthesis and Orichalcum Pulse) 

## Example
melonDS_20260201-152106338

melonDS_20260207-205701824

### CODE REVIEW

File: include/battle.h [hunk 1]
```diff
@@ -405,6 +405,7 @@
 #define WEATHER_SANDSTORM_ANY               (WEATHER_SANDSTORM | WEATHER_SANDSTORM_PERMANENT)                               // 0000 0000 0000 0000 1100
 #define WEATHER_SUNNY                       (0x00000010)                                                                    // 0000 0000 0000 0001 0000
 #define WEATHER_SUNNY_PERMANENT             (0x00000020)                                                                    // 0000 0000 0000 0010 0000
+#define WEATHER_SUNNY_NOT_EXTREMELY_HARSH   (WEATHER_SUNNY | WEATHER_SUNNY_PERMANENT)                                       // 0000 0000 0000 0011 0000 because protosynthesis
```
Comment (Koekenpann) [inline include/battle.h:409]: Desolate Land is in Scarlet and Violet, but only in a hacked state. Because Primal Groudon is not in the game, the ability is really just inaccessible/dexited. However, Desolate Land does not activate Protosynthesis which is the current implementation. I'm fine with it activating Protosynthesis.

Video: https://www.youtube.com/watch?v=OJO9TAIebro

File: data/battle_scripts/subscripts/subscript_0362_HANDLE_RAIN_TEMPORARY.s [hunk 2]
```diff
@@ -11,6 +11,9 @@ _000:
     UpdateVar OPCODE_FLAG_OFF, BSCRIPT_VAR_FIELD_CONDITION, FIELD_CONDITION_WEATHER
     UpdateVar OPCODE_FLAG_ON, BSCRIPT_VAR_FIELD_CONDITION, FIELD_CONDITION_RAIN
     UpdateVar OPCODE_SET, BSCRIPT_VAR_WEATHER_TURNS, 5
+    ResetParadoxAbility ABILITY_PROTOSYNTHESIS
+    // Booster Energy
+    ActivateParadoxAbility ABILITY_PROTOSYNTHESIS
```
Comment (Koekenpann) [inline data/battle_scripts/subscripts/subscript_0362_HANDLE_RAIN_TEMPORARY.s:16]: This will also make sure Booster Energy + Protosynthesis activate as soon as a non-Sun weather is setup.

Interaction: https://bsky.app/profile/nerdofnow.bsky.social/post/3lg2ox2cdlc2v

Comment (Koekenpann) [reply data/battle_scripts/subscripts/subscript_0362_HANDLE_RAIN_TEMPORARY.s:16]: I have also tested whether this applies to overworld Rain or not and it doesn't.
It goes Intimidate, then Booster + Protosynthesis.

So these commands aren't implemented in the overworld equivalent.

Image

Image

File: data/battle_tests/abilities/protosynthesis/activate_after_non_sun_weather.c
Comment (Koekenpann) [inline data/battle_tests/abilities/protosynthesis/activate_after_non_sun_weather.c:1]: Not sure how many tests is too many or enough.

I could definitely combine some of these though or add more if needed.

File: src/individual/SwitchInAbilityCheck.c [hunk 3]
```diff
@@ -733,6 +733,34 @@ int UNUSED SwitchInAbilityCheck(void *bw, struct BattleStruct *sp)
...
+                        }
+                    }
+
+                    // Hadron Engine
+                    {
+                        if ((sp->battlemon[client_no].hp)
+                        && (sp->battlemon[client_no].ability_activated_flag == 0)
+                        && (GetBattlerAbility(sp, client_no) == ABILITY_HADRON_ENGINE)) {
+                            sp->battlemon[client_no].ability_activated_flag = 1;
+                            sp->ability_client = client_no; // Use ability_client instead of battlerIdTemp so ActivateParadoxAbility doesn't interfere
```
Comment (Koekenpann) [inline src/individual/SwitchInAbilityCheck.c]: I don't think it actually interferes with Hadron Engine because Quark Drive message doesn't go inbetween two messages, unlike Orichalcum Pulse. Just to stay consistent with Orichalcum Pulse really.

File: include/battle.h (hunk 1 above)
Comment (BluRosie) [reply include/battle.h:409]: i think it would be kinda weird not to have it activate it as a sunny weather version.  but i can also understand the desire for purity
